torch_np/_detail/_util.py

```diff
@@ -0,0 +1,299 @@
+    return axis
+
+
+def allow_only_single_axis(axis):
+    if axis is None:
+        return axis
+    if len(axis) != 1:
+        raise NotImplementedError("does not handle tuple axis")
+    return axis[0]
+
+
+def expand_shape(arr_shape, axis):
+    # taken from numpy 1.23.x, expand_dims function
+    if type(axis) not in (list, tuple):
+        axis = (axis,)
+    out_ndim = len(axis) + len(arr_shape)
+    axis = normalize_axis_tuple(axis, out_ndim)
+    shape_it = iter(arr_shape)
+    shape = [1 if ax in axis else next(shape_it) for ax in range(out_ndim)]
+    return shape
+
+
+def apply_keepdims(tensor, axis, ndim):
+    if axis is None:
+        # tensor was a scalar
+        tensor = torch.full((1,)*ndim, fill_value=tensor, dtype=tensor.dtype)
+    else:
+        shape = expand_shape(tensor.shape, axis)
+        tensor = tensor.reshape(shape)
+    return tensor
+
+
+def axis_none_ravel(*tensors, axis=None):
+    """Ravel the arrays if axis is none."""
+    # XXX: is only used at `concatenate`. Inline unless reused more widely
+    if axis is None:
+        tensors = tuple(ar.ravel() for ar in tensors)
+        return tensors, 0
+    else:
+        return tensors, axis
+
+
+def cast_dont_broadcast(tensors, target_dtype, casting):
+    """Dtype-cast tensors to target_dtype.
+
+    Parameters
+    ----------
+    tensors : iterable
+	    tuple or list of torch.Tensors to typecast
+    target_dtype : torch dtype object, optional
+        The array dtype to cast all tensors to
+    casting : str
+        The casting mode, see `np.can_cast`
+
+    Returns
+    -------
+    a tuple of torch.Tensors with dtype being the PyTorch counterpart
+    of the `target_dtype`
+    """
+    # check if we can dtype-cast all arguments
+    cast_tensors = []
+    can_cast = _scalar_types._can_cast_impl
+
+    for tensor in tensors:
+        if not can_cast(tensor.dtype, target_dtype, casting=casting):
+            raise TypeError(f"Cannot cast array data from {tensor.dtype} to"
+                            f" {target_dtype} according to the rule '{casting}'")
+
+        # cast if needed
+        if tensor.dtype != target_dtype:
+            tensor = tensor.to(target_dtype)
+        cast_tensors.append(tensor)
+
+    return tuple(cast_tensors)
+
+
+def cast_and_broadcast(tensors, out_param, casting):
+    """
+    Parameters
+    ----------
+    tensors : iterable
+	tuple or list of torch.Tensors to broadcast/typecast
+    target_dtype : a torch.dtype object
+        The torch dtype to cast all tensors to
+    target_shape : tuple
+        The tensor shape to broadcast all `tensors` to
+    casting : str
+        The casting mode, see `np.can_cast`
+
+    Returns
+    -------
+    a tuple of torch.Tensors with dtype being the PyTorch counterpart
+    of the `target_dtype` and `target_shape`
+    """
+    if out_param is None:
+        return tensors
+
+    target_dtype, target_shape = out_param
+
+    can_cast = _scalar_types._can_cast_impl
```

Comment:
Does this mean that a better place for this function would be the file `_scalar_types.py`?

Reply:
The current logic is to keep `cast_and_broadcast` and `cast_dont_broadcast` next to each other (one is used by ufuncs, the other is used by `concatenate` et al). And wanted to keep `_scalar_types` minimal. But this can go either way, do you want me to move it?

Reply:
up to you, I'm fine with either.